File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/auth/refresh/RefreshTokenControllerTest.java
package br.com.deltaglobalbank.identity.features.auth.refresh;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import br.com.deltaglobalbank.identity.domain.token.InvalidRefreshTokenException;
import br.com.deltaglobalbank.identity.infrastructure.security.cookie.RefreshTokenCookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

class RefreshTokenControllerTest {

    private final RefreshTokenUseCase useCase = mock(RefreshTokenUseCase.class);
    private final RefreshTokenCookie refreshCookie = mock(RefreshTokenCookie.class);
    private final RefreshTokenController controller = new RefreshTokenController(useCase, refreshCookie);

    private final HttpServletRequest httpRequest = mock(HttpServletRequest.class);
    private final HttpServletResponse httpResponse = mock(HttpServletResponse.class);

    @Test
    void shouldReadTokenFromCookieAndRotateCookieOnRefresh() {
        when(refreshCookie.read(httpRequest)).thenReturn("cookie-token");
        when(useCase.execute(eq("cookie-token"), any()))
            .thenReturn(new RefreshTokenResult("access", "refresh", 900, false));

        ResponseEntity<RefreshTokenResponse> response = controller.refreshToken(null, httpRequest, httpResponse);

        verify(useCase, times(1)).execute(eq("cookie-token"), any());
        verify(refreshCookie, times(1)).set(httpResponse, "refresh");
        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertEquals("access", response.getBody().accessToken());
    }

    @Test
    void shouldFallBackToBodyTokenWhenCookieIsAbsent() {
        when(refreshCookie.read(httpRequest)).thenReturn(null);
        when(useCase.execute(eq("body-token"), any()))
            .thenReturn(new RefreshTokenResult("access", "refresh", 900, false));

        controller.refreshToken(new RefreshTokenRequest("body-token"), httpRequest, httpResponse);

        verify(useCase, times(1)).execute(eq("body-token"), any());
    }

    @Test
    void shouldThrowWhenNeitherCookieNorBodyHasAToken() {
        when(refreshCookie.read(httpRequest)).thenReturn(null);

        assertThrows(InvalidRefreshTokenException.class, () -> controller.refreshToken(null, httpRequest, httpResponse));
        verify(useCase, times(0)).execute(any(), any());
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/auth/refresh/RefreshTokenControllerTest.kt
package br.com.deltaglobalbank.identity.features.auth.refresh

import br.com.deltaglobalbank.identity.domain.token.InvalidRefreshTokenException
import br.com.deltaglobalbank.identity.infrastructure.security.cookie.RefreshTokenCookie
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.http.HttpStatus
import kotlin.test.Test
import kotlin.test.assertEquals

@ExtendWith(MockKExtension::class)
class RefreshTokenControllerTest {

    private val useCase = mockk<RefreshTokenUseCase>()
    private val refreshCookie = mockk<RefreshTokenCookie>(relaxed = true)
    private val controller = RefreshTokenController(useCase, refreshCookie)

    private val httpRequest = mockk<HttpServletRequest>(relaxed = true)
    private val httpResponse = mockk<HttpServletResponse>(relaxed = true)

    @Test
    fun `should read token from cookie and rotate cookie on refresh`(){
        every { refreshCookie.read(httpRequest) } returns "cookie-token"
        every { useCase.execute("cookie-token", any()) } returns RefreshTokenResult(accessToken = "access", refreshToken = "refresh", expiresIn=900, mustChangePassword=false)

        val response = controller.refreshToken(null, httpRequest, httpResponse)

        verify { useCase.execute("cookie-token", any()) }
        verify { refreshCookie.set(httpResponse, "refresh") }
        assertEquals(HttpStatus.OK, response.statusCode)
        assertEquals("access", response.body!!.accessToken)
    }

    @Test
    fun `should fall back to body token when cookie is absent`(){
        every { refreshCookie.read(httpRequest) } returns null
        every { useCase.execute("body-token", any()) } returns
                RefreshTokenResult("access", "refresh", expiresIn = 900, mustChangePassword = false)

        controller.refreshToken(RefreshTokenRequest("body-token"), httpRequest, httpResponse)

        verify { useCase.execute("body-token", any()) }
    }

    @Test
    fun `should throw when neither cookie nor body has a token`(){
        every { refreshCookie.read(httpRequest) } returns null

        assertThrows(InvalidRefreshTokenException::class.java) {
            controller.refreshToken(null, httpRequest, httpResponse)
        }
        verify(exactly = 0) { useCase.execute(any(), any()) }
    }

}